test/metabase/api/card_test.clj

```diff
@@ -3847,3 +3847,56 @@
                                        false
                                        (apply original-can-read? args)))]
           (is (map? (mt/user-http-request :crowberto :get 200 (format "card/%d/query_metadata" (:id card))))))))))
+
+(deftest dashboard-internal-card-creation
+  (mt/with-temp [:model/Collection {coll-id :id} {}
+                 :model/Collection {other-coll-id :id} {}
+                 :model/Dashboard {dash-id :id} {:collection_id coll-id}]
+    (testing "We can create a dashboard internal card"
+      (mt/user-http-request :crowberto :post 200 "card" (assoc (card-with-name-and-query)
+                                                               :dashboard_id dash-id)))
+    (testing "We can't create a dashboard internal card with a collection_id different that its dashboard's"
+      (mt/user-http-request :crowberto :post 400 "card" (assoc (card-with-name-and-query)
+                                                               :dashboard_id dash-id
+                                                               :collection_id other-coll-id)))
+    (testing "Note the bug: if you provide `collection_id=null`, we'll accept it."
```

Comment:
Wdym by bug? Is this something that should be changed, or expected behavior?

Reply:
Ah, good callout. I had meant to come back to this.

Normally if you provide a `collection_id` we require that it's the same as the `collection_id` of the dashboard you're telling us to put the card in, and respond with a 4XX otherwise.

But in the case where you provide `collection_id=null`, and the dashboard _isn't_ in the root collection, we ignore the mismatch. We don't _do_ anything crazy - we just put the card into the dashboard's collection, not the root collection, and ignore the `collection_id` you passed in (it's effectively overridden by the `dashboard_id`).

The reason for this is that by the time we're in the before-insert toucan hook, we've set any missing values to `nil` in `metabase.models.card/create-card!` - but I've just looked it over and I think I can avoid that pretty easily. We were just doing that for the purpose of calling `api/maybe-reconcile-collection-position!`, which needs `collection_id` and `collection_position` to be present. I'll just rework things to keep the card-data the same (with missing keys) for `t2/insert!` and call `api/maybe-reconcile-collection-position!` with a different map. Then this bug will be gone and things should be consistent.